File: src/task12_3_1/Engine.java
package task12_3_1;

public class Engine {
    public enum Fuel {
        Petrol,
        Diesel,
        Electric
    }

    final int power;
    final Fuel fuel;

    public Engine(int power, Fuel fuel) {
        this.power = power;
        this.fuel = fuel;
    }

    @Override
    public String toString() {
        return "Engine{" + this.hashCode() +
                " power=" + power +
                ", fuel=" + fuel +
                '}';
    }
}

File: src/task6_3_2/Engine.kt
package task6_3_2


class Engine: Cloneable {

    enum class EngineType {
        DIESEL,
        HYBRID,
        ROTARY,
        GASOLINE
    }

    class Builder{

        private var engineType: EngineType = EngineType.DIESEL

        fun setEngineType(engineType: EngineType): Builder{
            this.engineType = engineType
            return this
        }

        fun build(): Engine{
            return Engine(engineType)
        }

    }

    private val engineType: EngineType

    private constructor(engineType: EngineType){
        this.engineType = engineType
    }

    override fun toString(): String {
        return "${this.hashCode()}-Engine - type: $engineType\n"
    }

    override fun copy(): Engine{
        return Engine(this.engineType)
    }

}